meta-mbl-distro/scripts/lib/update/images.py

```diff
@@ -0,0 +1,331 @@
+                    AppsImage._get_control_tgz_from_ipk(ipk, ipk_path),
+                    ipk_path,
+                ),
+                ipk_path,
+            )
+
+    @staticmethod
+    def _get_control_tgz_from_ipk(ipk, ipk_path):
+        """
+        Get the "control.tar.gz" file from within an ipk.
+
+        Args:
+        * ipk file-like: an open ipk file.
+        * ipk_path path-like: the path to the ipk (for error messages).
+        """
+        # ipk files are "ar" archives
+        ar = arpy.Archive(fileobj=ipk)
+        ar.read_all_headers()
+        control_tgz_fname = b"control.tar.gz"
+        if control_tgz_fname not in ar.archived_files:
+            bb.fatal(
+                'Failed  to find "{}" file in app file "{}"'.format(
+                    control_tgz_fname, ipk_path
+                )
+            )
+        return ar.archived_files[control_tgz_fname]
+
+    @staticmethod
+    def _get_control_file_from_control_tgz(control_tgz, ipk_path):
+        """
+        Get the "control" file from the "control.tar.gz" from an ipk.
+
+        Args:
+        * control_tgz file-like: an open control.tar.gz file.
+        * ipk_path: the path to the ipk (for error messages).
+        """
+        tar = tarfile.open(fileobj=control_tgz, mode="r|gz")
+        for tarinfo in tar:
+            if not tarinfo.name.endswith("control"):
+                continue
+            return tar.extractfile(tarinfo)
+        bb.fatal(
+            'Failed to find "control" file in app file "{}"'.format(ipk_path)
+        )
+
+    @staticmethod
+    def _get_app_name_from_control_file(control_file, ipk_path):
+        """
+        Get the app name from the "control" file from an ipk.
+
+        Args:
+        * control_file file-like: an open control file.
+        * ipk_path: the path to the ipk (for error messages).
+        """
+        for line in control_file.read().decode("utf-8").splitlines():
+            field, value = line.split(sep=":", maxsplit=1)
+            if field.strip() == "Package":
+                return value.strip()
+        bb.fatal(
+            'Failed find app name in "control" file from "{}"'.format(ipk_path)
+        )
+
+
+class BootImage:
+    def __init__(self, deploy_dir, tinfoil):
+        self._boot_files = self._get_archived_file_specs(deploy_dir, tinfoil)
+        self.image_type = MBL_BOOT_ID
+        self.image_format_version = 3
+        self.archived_path = pathlib.Path("{}.tar.xz".format(MBL_BOOT_ID))
+
+    def stage(self, staging_dir):
+        _stage_multi_file_component(
+            staging_dir, self.archived_path, self._boot_files
+        )
+
+    def generate_testinfo(self):
+        return [testinfo.file_compare("/proc/version")] + [
+            testinfo.file_sha256(
+                afs.path, pathlib.Path("/boot", afs.archived_path)
+            )
+            for afs in self._boot_files
+        ]
+
+    @staticmethod
+    def _get_archived_file_specs(deploy_dir, tinfoil):
```

Comment:
> I guess you could argue this method isn't changing the object's state so it should be static (or a free function)

Exactly - the method doesn't access any object state. Marking it as static means that readers immediately know that, so understanding the class, and the method itself, should be easier.

>  but I'm still not sure I like this shape.

Why?

Reply:
I think I'd prefer we had another module with free functions rather than all these `staticmethods`; but as I said in the RL discussion, I'm not going to gate the PR on it, as it's my stylistic preference rather than having any technical argument for changing it.

Reply:
Addressed in https://github.com/ARMmbed/meta-mbl/pull/763/commits/ecb7eeeaacee2e09b5f38ef595037176c4fe2b5a